### ISSUE TITLE
Twisted crashes when parsing XMPP XML stream with Python 3.8

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#9730]
|[<img alt="goffi-contrib's avatar" src="https://avatars.githubusercontent.com/u/22869331?s=50" width="50" height="50">](https://github.com/goffi-contrib)| @goffi-contrib reported|
|-|-|
|Trac ID|trac#9730|
|Type|defect|
|Created|2019-11-17 17:15:22Z|

Hello,

When parsing an XML stream with Python 3.8, `twisted.words.xish.domish import ExpatElementStream` is crashing with the following traceback:

```python
        Traceback (most recent call last):
          File "/home/goffi/venvs/sat_principal/lib/python3.8/site-packages/twisted/python/log.py", line 103, in callWithLogger
            return callWithContext({"system": lp}, func, *args, **kw)
          File "/home/goffi/venvs/sat_principal/lib/python3.8/site-packages/twisted/python/log.py", line 86, in callWithContext
            return context.call({ILogContext: newCtx}, func, *args, **kw)
          File "/home/goffi/venvs/sat_principal/lib/python3.8/site-packages/twisted/python/context.py", line 122, in callWithContext
            return self.currentContext().callWithContext(ctx, func, *args, **kw)
          File "/home/goffi/venvs/sat_principal/lib/python3.8/site-packages/twisted/python/context.py", line 85, in callWithContext
            return func(*args,**kw)
        --- <exception caught here> ---
          File "/home/goffi/venvs/sat_principal/lib/python3.8/site-packages/twisted/internet/posixbase.py", line 614, in _doReadOrWrite
            why = selectable.doRead()
          File "/home/goffi/venvs/sat_principal/lib/python3.8/site-packages/twisted/internet/tcp.py", line 243, in doRead
            return self._dataReceived(data)
          File "/home/goffi/venvs/sat_principal/lib/python3.8/site-packages/twisted/internet/tcp.py", line 249, in _dataReceived
            rval = self.protocol.dataReceived(data)
          File "/home/goffi/venvs/sat_principal/lib/python3.8/site-packages/twisted/words/xish/xmlstream.py", line 93, in dataReceived
            self.stream.parse(data)
          File "/home/goffi/venvs/sat_principal/lib/python3.8/site-packages/twisted/words/xish/domish.py", line 797, in parse
            self.parser.Parse(buffer)
          File "/build/python/src/Python-3.8.0/Modules/pyexpat.c", line 407, in StartElement
            
          File "/home/goffi/venvs/sat_principal/lib/python3.8/site-packages/twisted/words/xish/domish.py", line 810, in _onStartElement
            for k, v in attrs.items():
        builtins.RuntimeError: dictionary keys changed during iteration
```

note that this was working as expected with Python 3.7.

I've made a small script which reproduce the issue, the value in buffer comes from real XMPP stream:

```python
#!/usr/bin/env python3.8

from twisted.words.xish.domish import ExpatElementStream


buf = (b"<?xml version='1.0'?><stream:stream from='sat-pubsub.tazar2.int' xmlns:stream='"
       b"http://etherx.jabber.org/streams' id='c2f2fecf-fb57-4a30-bbfe-0e9b5a587a8e' "
       b"xmlns='jabber:component:accept' xml:lang='en'>")

elem_stream = ExpatElementStream()
elem_stream.parse(buf)
print("test passed")
```

This makes Twisted unusable for XMPP with Python 3.8, so I set the priority as `highest`.

Thanks!

<details><summary>Searchable metadata</summary>

```
trac-id__9730 9730
type__defect defect
reporter__goffi_contrib goffi-contrib
priority__highest highest
milestone__Python_3_8_Support Python 3.8 Support
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__None None
time__1574010922876977 1574010922876977
changetime__1583042921916818 1583042921916818
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

Hi, thanks for the bug report. We don't officially support Python 3.8 yet, so I've added it to the 'support 3.8' milestone. I'd suggest using 3.7 for the time being :)

**Comment by twisted-trac** [comment]
|[<img alt="goffi-contrib's avatar" src="https://avatars.githubusercontent.com/u/22869331?s=50" width="50" height="50">](https://github.com/goffi-contrib)<a name="note_4"></a>|@goffi-contrib commented|
|-|-|

I've submitted a fix at https://github.com/twisted/twisted/pull/1198

**Comment by twisted-trac** [comment]
|[<img alt="goffi-contrib's avatar" src="https://avatars.githubusercontent.com/u/22869331?s=50" width="50" height="50">](https://github.com/goffi-contrib)<a name="note_6"></a>|@goffi-contrib commented|
|-|-|

Hello hawkowl, thanks for the feedback, I'll apply the fix in my own code until a Python 3.8 compatible version of Twisted is available, as it seems to be enough for me. I hope the PR I've made is fine.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @glyph|
|-|-|

Thanks for your contribution.  Looks good to me; I'll land it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set status to `closed`|
|-|-|

In changeset a24e9345c609f824a813febfc6af7f350812dbc0
```
#!CommitTicketReference repository="" revision="a24e9345c609f824a813febfc6af7f350812dbc0"
Merge pull request #1198 from goffi-contrib/9730-goffi-contrib-fix-xmpp-parsing-python-3.8

Author: goffi-contrib

Reviewer: glyph

Fixes: ticket:9730

Fix parsing of namespaced attributes with Python 3.8 in twisted.words.
```
